marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/flow/getFlowsWithStepDetails.sjs

```diff
@@ -71,6 +72,27 @@ flows.map(flow => {
       stepDetails.stepName = step.name;
       stepDetails.stepDefinitionType = step.stepDefinitionType;
       stepDetails.sourceFormat = step.sourceFormat;
+
+      const jobQueries = [];
+      jobQueries.push(cts.collectionQuery('Job'));
+      jobQueries.push(cts.jsonPropertyValueQuery("flow",flow.name));
+      jobQueries.push(cts.jsonPropertyValueQuery("stepName",step.name));
+
+      let latestJob = fn.head(datahub.hubUtils.queryLatest(function() {
```

Comment:
Pinging @ryanjdew and @fsnow  about this - having to do a query on every iteration of a loop (and this is a nested loop) is often a recipe for slow performance in production. Now, in theory, the query shouldn't happen that often - it would seem rare for a user to have more than a dozen flows, and for each of those flows to average a dozen or more steps. So if we're running this query 10 to 20 times on each call, it shouldn't add that much overhead. 

But, perhaps the performance is slightly worse because we're hitting a separate database. And we're also getting this job data every single time we get flows for the UI. 

I think before we go forward, we need to review the requirement to verify that it's worth doing this. The "when" in the scenario isn't valid - "When I look at the step in the Run tile" - that's not an event that the computer knows about. The way this implemented, it's assuming a "when" of "When I click on the Run tile" (or do something that causes the Run tile to be displayed). And is it true that we want job data every time this "when" happens? I don't know. 

@srinathgit  Can you organize a quick call with myself and Ryan and Justin to determine what the "when" should really be here? (No need to involve fsnow, just wanted to start getting him exposed to some backend code reviews)

Reply:
@srinathgit @rjrudin This one is difficult to reduce to one cts search given we need to get the latest job for each step. 

However, reducing this to a single queryLatest call which will reduce the number of invokes to another database should be doable. That would mean iterating over the steps twice. Once to get the job info in the queryLatest call and a second time for the additional step details and incorporating the job info. Despite the additional looping, the performance would be better in most scenarios taking that direction.

Reply:
We'd still have N queries against the jobs database, right? Where N = the number of steps across all the flows? 

The overhead of the eval should be much less than N queries. We'll use the meeting today to figure out if the requirements justify N queries - that's just been a common performance issue for me in the field - but it may be mitigated enough by the rarity of there being hundreds of steps in a DHF project.